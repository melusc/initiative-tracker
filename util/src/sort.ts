/*!
Copyright (C) Luca Schnellmann, 2025

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

// Reminder:
//	Return -1 for a before b
//  Return 1 for b before a
//	Return 0 for equal

function isNullish(value: unknown): value is null | undefined {
	return value === null || value === undefined;
}

export function makeSorter<R extends Record<string, unknown>>(
	keys: Array<{key: keyof R; reverse: boolean}>,
) {
	return <T extends R>(array: T[]): T[] =>
		array.toSorted((a, b) => {
			for (const {key, reverse} of keys) {
				// Always nullish last
				// A before B
				if (!isNullish(a[key]) && isNullish(b[key])) {
					return -1;
				}

				// B before A
				if (isNullish(a[key]) && !isNullish(b[key])) {
					return 1;
				}

				// Go to next comparison
				if (isNullish(a[key]) && isNullish(b[key])) {
					continue;
				}

				const result = (a[key] as string).localeCompare(
					b[key] as string,
					undefined,
					{
						numeric: true,
						caseFirst: 'lower',
						sensitivity: 'base',
					},
				);

				if (result !== 0) {
					return reverse ? -result : result;
				}
			}

			return 0;
		});
}

export const sortInitiatives = makeSorter<{
	deadline: string | null | undefined;
	shortName: string | null | undefined;
	id: string;
}>([
	{key: 'deadline', reverse: true},
	{key: 'shortName', reverse: false},
	{key: 'id', reverse: false},
]);
export const sortPeople = makeSorter<{
	name: string | null | undefined;
	id: string;
}>([
	{key: 'name', reverse: false},
	{key: 'id', reverse: false},
]);
export const sortOrganisations = makeSorter<{
	name: string | null | undefined;
	id: string;
}>([
	{key: 'name', reverse: false},
	{key: 'id', reverse: false},
]);
