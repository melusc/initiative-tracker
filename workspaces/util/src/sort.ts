import type {Initiative, Organisation, Person} from './types.js';

type KeysMatching<T, V> = {
	[K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

function makeSorter<R extends Record<string, unknown>>(
	keys: Array<{key: KeysMatching<R, string>; reverse: boolean}>,
) {
	return (array: R[]) =>
		array.toSorted((a, b) => {
			for (const {key, reverse} of keys) {
				const result = (a[key] as string).localeCompare(
					b[key] as string,
					undefined,
					{
						numeric: true,
						caseFirst: 'lower',
						sensitivity: 'base',
					},
				);

				if (result !== 0) {
					return reverse ? -result : result;
				}
			}

			return 0;
		});
}

export const sortInitiatives = makeSorter<Initiative>([
	{key: 'deadline', reverse: true},
	{key: 'shortName', reverse: false},
	{key: 'id', reverse: false},
]);
export const sortPeople = makeSorter<Person>([
	{key: 'name', reverse: false},
	{key: 'id', reverse: false},
]);
export const sortOrganisations = makeSorter<Organisation>([
	{key: 'name', reverse: false},
	{key: 'id', reverse: false},
]);
