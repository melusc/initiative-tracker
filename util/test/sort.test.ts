/*!
Copyright (C) Luca Schnellmann, 2025

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import {expect, test} from 'vitest';

import {makeSorter} from '../src/sort.js';

const sorter = makeSorter<{
	a: string | null | undefined;
	b: string | null | undefined;
}>([
	{key: 'a', reverse: false},
	{key: 'b', reverse: false},
]);

const sorterReverse = makeSorter<{
	a: string | null | undefined;
	b: string | null | undefined;
}>([
	{key: 'a', reverse: true},
	{key: 'b', reverse: true},
]);

test.for([
	[
		{a: 'a', b: 'c'}, // 1
		{a: 'x', b: 'z'}, // 2
	],
	[
		{a: 'x', b: 'z'}, // 2
		{a: 'a', b: 'c'}, // 1
	],
	[
		{a: null, b: 'c'}, // 2
		{a: null, b: 'd'}, // 1
	],
	[
		{a: null, b: 'c'}, // 2
		{a: 'z', b: 'c'}, // 1
	],
	[
		{a: null, b: null}, // 2
		{a: null, b: 'a'}, // 1
	],
	[
		{a: undefined, b: undefined}, // 2
		{a: null, b: 'a'}, // 1
	],
])('Sorting logic', testCase => {
	const sorted = sorter([...testCase]);
	const sortedReverse = sorterReverse([...testCase]);

	expect(sorted).toMatchSnapshot();
	expect(sorted.toReversed()).toStrictEqual(sortedReverse);
});
