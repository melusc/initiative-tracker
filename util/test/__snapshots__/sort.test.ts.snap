// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sorting logic 1`] = `
[
  {
    "a": "a",
    "b": "c",
  },
  {
    "a": "x",
    "b": "z",
  },
]
`;

exports[`Sorting logic 2`] = `
[
  {
    "a": "a",
    "b": "c",
  },
  {
    "a": "x",
    "b": "z",
  },
]
`;

exports[`Sorting logic 3`] = `
[
  {
    "a": null,
    "b": "c",
  },
  {
    "a": null,
    "b": "d",
  },
]
`;

exports[`Sorting logic 4`] = `
[
  {
    "a": "z",
    "b": "c",
  },
  {
    "a": null,
    "b": "c",
  },
]
`;

exports[`Sorting logic 5`] = `
[
  {
    "a": null,
    "b": "a",
  },
  {
    "a": null,
    "b": null,
  },
]
`;

exports[`Sorting logic 6`] = `
[
  {
    "a": null,
    "b": "a",
  },
  {
    "a": undefined,
    "b": undefined,
  },
]
`;
